sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -62,6 +61,18 @@ object TypeCoercion {
       WindowFrameCoercion ::
       Nil
 
+  def rules(conf: SQLConf): List[Rule[LogicalPlan]] = {
+    if (conf.isHiveTypeCoercionMode) {
+      commonTypeCoercionRules :+
+        HiveInConversion :+
+        HivePromoteStrings
+    } else {
+      commonTypeCoercionRules :+
+        InConversion :+
+        PromoteStrings
```

Comment:
These two rules are moved to the end of the whole batch. It should be fine after https://github.com/apache/spark/pull/19867.